### PR TITLE
Fix stage/task leak

### LINKED ISSUE #2387: Tasks continue running even after query has finished
Coordinator is calling getStatus on tasks after the query has finished. This keeps the tasks alive forever.

### PR DESCRIPTION
**Pull request by cberner**
Fixes #2387

### CODE REVIEW

**Comment by cberner** [comment]
@frsyuki I think this might fix an issue that you guys were seeing

File: presto-docs/src/main/sphinx/release/release-0.95.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+============
+Release 0.95
+============
+
+General Changes
+---------------
+
+* Fix task and stage leak, caused when a stage finishes before its substages
```
Comment (dain) [inline presto-docs/src/main/sphinx/release/release-0.95.rst]: verify the tense of this compared to the other release notes

**Comment by dain** [comment]
Looks good

**Comment by frsyuki** [comment]
@cberner Thank you for letting us know! We'll try.
Does it cause this scenario?:

1. at a parent stage, all tasks started but failed because all workers returned non-successful HTTP response (due to crash of the worker processes, timeout, etc.)
2. `doUpdateState()` changes state of the parent stage to FAILED, which means done.
3. if some substages are still running at this moment, `cancel()` call at the end of `doUpdateState()` doesn't cancel the running substages.
4. because the parent tasks of the running stages don't pull buffers, substages are stalled.

Does here also need `cancel()` call?:
https://github.com/cberner/presto/blob/cancel/presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java#L882

**Comment by cberner** [comment]
Yes, that's exactly the situation it fixes. Another situation that can happen is that the parent stage finishes because it contains a LimitOperator, even if there are no errors.

I don't think we need a `cancel()` there, since `cancel()` or `abort()` is called after every transition. I'll ask @dain to confirm though

**Comment by frsyuki** [comment]
I see. thank you!

**Pull request merged**